### ISSUE TITLE
Store change listener fired before dispatcher's finished dispatching.

### ISSUE THREAD

**Issue by wayneashleyberry** [facebook/flux#138]
This is seemingly related to all of those "Cannot dispatch in the middle of a dispatch" issues, but I've narrowed it down to a store listener being triggered _before_ the dispatcher's finished up.

Should that ever happen?

I can use the `setTimeout` hack, or the queue suggested in #106 but it doesn't feel right.

In my setup I have a login component with the following listener:

```js
  onAuthStoreChange: function() {
    if (AuthStore.userIsAuthenticated() === true) {
      this.transitionTo('people');
    }
  },
```

Now when that gets fired I kind of expect the dispatcher to be finished dispatching... but it's not. So then when I transition to a new route and new actions are created there are obviously problems...

Is there any way to get listeners to only be triggered once the dispatcher's done?

**Comment by wayneashleyberry** [comment]
p.s. Hope everyone's having fun at the conf :)

**Comment by sterpe** [comment]
I think this is expected.  The stores' dispatcher-registered listeners are called sequentially on a trampoline of sorts, so the #dispatch() function call isn't popped off the stack until all of the stores's callbacks have returned, then #dispatch() returns.  Your store is likely firing it's 'change' event in the same execution context, so `onAuthStoreChange` is called while the function call to #dispatch() is still waiting on the stack waiting to return.

Afaik, the only thing you're guaranteed is that any required browser repaint won't happen until after the call to #dispatch() returns, and that that browser repaint will include all updates to reflect the a singular point in time/state that exists after the last dispatch.

**Comment by fisherwebdev** [comment]
I don't think we are going to change Dispatcher to provide a hook *after* the dispatch completes, but internally we are looking very seriously at updating Dispatcher to provide a hook after all data mutations and just before `_stopDispatching` is called.  That is, just before this line: https://github.com/facebook/flux/blob/master/src/Dispatcher.js#L195  That change will make it GitHub if we find that it helps us internally at FB with our applications.

**Comment by fantasy8-me** [comment]
I encounter similar error.
`Error: Invariant Violation: Dispatch.dispatch(...): Cannot dispatch in the middle of a dispatch`

I got this error when I tried to start another dispatch in a `listener` that is triggered by `emitChange()` of a `Store`. I can workaround this by below two approaches, but both look like a hack

1. Invoke listener method by setTimeout
1. Make the `emit` async (using setTimeout as well)

  ```
  emitChange: function() {
    var self = this;
    window.setTimeout(function() {
      self.emit(CHANGE_EVENT);
    }, 1)
  },
  ```

So my question is that when a dispatch should be treated as done/completed, after the executions of all listeners on a `Store', or not care the execution of the listener (in which case I think we should make the emit async)

**Comment by fisherwebdev** [comment]
^^^ That's a hack that should be avoided.  Dispatch-within-a-dispatch errors are caused by improper code design 99.9% of the time.  Instead of doing the setTimeout, do what you need to do in response to the original action.  I think this is most often a reflection of equating actions with setters -- they are not the same.

**Comment by sterpe** [comment]
@fisherwebdev how do you guys feel about serial dispatches inside some Actioncreator.action()?

I.e: 

dispatcher.dispatch({});
dispatcher.dispatch({});

Is this ok or better to merge into a single action type?

**Comment by LuisHerranz** [comment]
@sterpe that's an interesting question. Do you have a real example?

I guess you should fire only one action which describes in whole what happened and stores should react to that single change.

I want to know what @fisherwebdev has to say anyway.

**Comment by sterpe** [comment]
Say WebAPIUtils does something like 

ContactsApiGetsResponse.then( f(x) {
// inform spinner status store or whatever emit change
dispatcher.dispatch ({
actiontype: "response-succes"
data:null
})
//store should load this new data and emit change
dispatcher.dispatch({
actiontype: "load-API-data-into - relevant store"
data:x.responsedata
})

**Comment by LuisHerranz** [comment]
In that case I think the store which is interested on that data should listen to the "response-success" event.

**Comment by sterpe** [comment]
Point being that the spinner store is really only a Boolean toggle and seems like it shouldn't need cases for every API actiontype just a generic request-started requested-ended : put up and pull down the spinner modal?

**Comment by LuisHerranz** [comment]
As @fisherwebdev has said above, you shouldn't confuse actions with setters.

Maybe you can do a regex of the actionType and check if it has "API" and "success" in it, or something like that.

**Comment by sterpe** [comment]
@LuisHerranz, In this case I don't feel like there's a confusion of setters and actions, though I can stand to be corrected.  When my data store receives new items that is a server initiated action -- load these new items into the store and by extension dependent stores (through waitFor on same action) and component views respond to the emitted change.  Otherwise it seems like you're proposing calling the store.loadItems method directly from say WebAPIUtils, skipping the dispatcher entirely, and then force emitting a change event so dependent views would rerender.  I'm not sure how you see a store hierarchy working in that case?

In my particular case I must float a div at top of page on every CRUD action indicating that the changes were saved (or conversely not saved).  This component (and it's store) doesn't particularly care about the details of any particular api request WebAPIUtils makes behalf of a particular data store that the user is manipulating.  So not sure if it makes sense for this component to be aware of the unique actions that are relevant to unique stores when WebAPIUtils makes requests.

**Comment by sterpe** [comment]
@LuisHerranz, I don't know...I can't buy into regexing actionType strings, that seems potentially very brittle.

**Comment by LuisHerranz** [comment]
What I mean by confusing setters and actions is that you shouldn't name your actions like this:

``actiontype: "load-API-data-into - relevant store"```

You should name them like

```actiontype: "relevant-data-received"```

and the relevant store, which is interested in that data, should be listening to that action.

**Comment by sterpe** [comment]
Perhaps, but that's a semantic issue in action naming--I feel like the heart is the question of what point you collapse actions together vs perform sequential dispatches, b/c say a certain component needs to respond to a multitude of different actions but in some sense doesn't need to be aware of them.

**Comment by sterpe** [comment]
*component should be 'store'

**Comment by vladar** [comment]
@fisherwebdev There is another problem with firing emitChange in the middle of dispatch process: your app may receive inconsistent state from stores.

Consider simple example: Component depends on stores "A" & "B". Store "B" waitsFor store "A". When store "A" receives action and does "emitChange" store "B" hasn't done yet. Your component tries to update and recieves new state from store "A", but old state from store "B".

I faced with this issue with item / list stores, but pretty sure there might be other scenarios as well.

The only good way to solve this is to emitChange after all stores are done mutating. But there is no such option in Dispatcher yet. So what else can you do except setTimeout(emitChange, 0) at this point?

**Comment by fisherwebdev** [comment]
@vladar There are a couple of different strategies to solving this problem. We believe we have a good solution and hope to open source it soon.  But it's not the only solution.

Options:

1. Debounce the `change` events in the React components.  We currently think this is the best solution.
2. Modify the dispatcher to call another set of callbacks at the end of the dispatch cycle.  We went pretty far down this road, but in the end it did not seem as good as debouncing the change events.  I think the primary concern was that it didn't seem like something the Dispatcher should be responsible for.  It seemed like we were putting code in the Dispatcher that didn't really have anything to do with its original purpose, but was really tied to the store-view communication.
3. Create an "OmegaStore" -- one that will `waitFor` all the other stores to update, and then emit change if those stores have changed.  This solution seemed cool at first, but then wound up looking like a gigantic hack.

**Comment by admmasters** [comment]
Out of interest, is debouncing the change events in the components the solution you are open sourcing, or is there another solution on its way?

**Comment by LuisHerranz** [comment]
You can use underscore, for example:
http://underscorejs.org/#debounce

**Comment by admmasters** [comment]
Yeah we are using lodash - I meant in terms of the wording, wanted to know whether or not debouncing was what was being referred to in the following: "We believe we have a good solution and hope to open source it soon", as it implicitly suggested there was another solution not including the options listed.

**Comment by ekryski** [comment]
Man, still feels like a bit of a hack considering there is already `waitFor()`. Sort of feels like the `waitFor` method should do this for you.

Regardless, debouncing does work quite well. For anyone that is interested here is what I ended up doing:

```js
var _ = require('lodash');
var CHANGE_EVENT = "change";
var ApplicationStore = _.extend({}, EventEmitter.prototype, {
  
  emitChange: function() {
    this.emit(CHANGE_EVENT);
  },

  addChangeListener: function(callback) {
    this.on(CHANGE_EVENT, _.debounce(callback, 100));
  },

  removeChangeListener: function(callback) {
    this.removeListener(CHANGE_EVENT, callback);
  },

  getState: function(){
    return {
      showConsole: _showConsole
    }
  }
});
```

**Comment by ekryski** [comment]
As a little update I ran into an issue using `debounce`. If you don't call `.cancel()` on the debounced function then you'll end up with event leaks because the `removeListener` doesn't actually remove the callback.

**Comment by facebook-github-bot-6** [comment]
Thank you for reporting this issue and appreciate your patience. We've  notified the core team for an update on this issue. We're looking for a response within the next 30 days or the issue may be closed.

**Comment by kyldvs** [comment]
Just pushed 2.1.0 which releases the Mixin/Container classes we use to solve the debounce issue.

https://github.com/facebook/flux/blob/master/src/FluxStoreGroup.js Allows you to add a hook after the dispatches of certain stores have finishes (but before the dispatch is done)

https://github.com/facebook/flux/blob/master/src/FluxContainer.js#L89 This is how we use it to synchronously "debounce" calls to setState in a react component.

**Comment by backslash112** [comment]
@ekryski How to call the cancel() method on the function?
like this:
```
this.method = _. debounce(callback, 100) 
this.method.cancel()?
```
It's doesn't work... ;(

**Comment by backslash112** [comment]
call .cancel() on the debounced function like this:
[http://stackoverflow.com/a/34830349/2195426](http://stackoverflow.com/a/34830349/2195426)

**Comment by MarosPixel** [comment]
Hack with setTimeout in emit worked for me.

```javascript
import { EventEmitter } from "events";

export default class Store extends EventEmitter
{
    emitChange()
    {
        // this.emit('change')
        setTimeout(() => this.emit('change'), 0);
    }

    addChangeListener(callback)
    {
        this.on('change', callback);
    }

    removeChangeListener(callback)
    {
        this.removeListener('change', callback);
    }
}
```
